File: it-agents/skills/code-architect/references/code-examples/cpp/Memento.cpp
// Memento — capture and restore an object's state without breaking encapsulation.
#include <iostream>
#include <stack>
#include <string>

// Memento stores an opaque snapshot; only the Originator reads it.
class Memento {
    friend class Editor;
    std::string state;
    explicit Memento(std::string s) : state(std::move(s)) {}
};

// Originator.
class Editor {
    std::string buffer;
public:
    void type(const std::string& s) { buffer += s; }
    const std::string& text() const { return buffer; }

    Memento save() const { return Memento{buffer}; }
    void restore(const Memento& m) { buffer = m.state; }
};

// Caretaker keeps mementos without inspecting them.
class History {
    std::stack<Memento> snapshots;
public:
    void push(const Memento& m) { snapshots.push(m); }
    Memento pop() {
        Memento m = snapshots.top();
        snapshots.pop();
        return m;
    }
    bool empty() const { return snapshots.empty(); }
};

int main() {
    Editor editor;
    History history;

    editor.type("alpha");
    history.push(editor.save());
    editor.type("-beta");
    std::cout << "current: " << editor.text() << "\n";

    editor.restore(history.pop());
    std::cout << "restored: " << editor.text() << "\n";
    return 0;
}

File: it-agents/skills/code-architect/references/code-examples/swift/Memento.swift
// Memento — capture and restore an object's state without breaking encapsulation.

// Opaque snapshot; only the originator knows how to read it.
struct EditorMemento {
    fileprivate let content: String
    fileprivate let cursor: Int
}

final class Editor {
    private var content = ""
    private var cursor = 0

    func type(_ text: String) {
        content += text
        cursor = content.count
    }

    func snapshot() -> EditorMemento {
        EditorMemento(content: content, cursor: cursor)
    }

    func restore(_ memento: EditorMemento) {
        content = memento.content
        cursor = memento.cursor
    }

    var state: String { "\"\(content)\" (cursor: \(cursor))" }
}

// Caretaker stores mementos without inspecting them.
final class History {
    private var stack: [EditorMemento] = []
    func save(_ memento: EditorMemento) { stack.append(memento) }
    func pop() -> EditorMemento? { stack.popLast() }
}

func runDemo() {
    let editor = Editor()
    let history = History()

    editor.type("Hello")
    history.save(editor.snapshot())
    editor.type(", world")
    print("current:", editor.state)

    if let saved = history.pop() {
        editor.restore(saved)
    }
    print("restored:", editor.state)
}

runDemo()
